pkg/fleetautoscalers/fleetautoscalers_test.go

```diff
@@ -152,55 +79,550 @@ func (t testServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+				Service: &admregv1b.ServiceReference{
+					Name:      "service1",
+					Namespace: "default",
+					Path:      &url,
+				},
+				URL: &(server.URL),
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "service and url cannot be used simultaneously",
+			},
+		},
+		{
+			description: "URL not nil but empty",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: nil,
+				URL:     &emptyString,
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "URL was not provided",
+			},
+		},
+		{
+			description: "Invalid URL",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: nil,
+				URL:     &invalidURL,
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "parse )1golang.org/: invalid URI for request",
+			},
+		},
+		{
+			description: "Service name is empty",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: &admregv1b.ServiceReference{
+					Name:      "",
+					Namespace: "default",
+					Path:      &url,
+				},
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "service name was not provided",
+			},
+		},
+		{
+			description: "No certs",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: &admregv1b.ServiceReference{
+					Name:      "service1",
+					Namespace: "default",
+					Path:      &url,
+				},
+				CABundle: []byte("invalid-value"),
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "no certs were appended from caBundle",
+			},
+		},
+		{
+			description: "Wrong server URL",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: nil,
+				URL:     &wrongServerURL,
+			},
+			expected: expected{
+				replicas: 0,
+				limited:  false,
+				err:      "Post http://127.0.0.1:1: dial tcp 127.0.0.1:1: connect: connection refused",
+			},
+		},
+		{
+			description: "Handle server error",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: nil,
+				URL:     &(server.URL),
+			},
+			specReplicas:   50,
+			statusReplicas: 50,
+			// hardcoded value in a server implementation
+			statusAllocatedReplicas: -10,
+			statusReadyReplicas:     40,
+			expected: expected{
+				replicas: 50,
+				limited:  false,
+				err:      fmt.Sprintf("bad status code %d from the server: %s", http.StatusInternalServerError, server.URL),
+			},
+		},
+		{
+			description: "Handle invalid response from the server",
+			webhookPolicy: &autoscalingv1.WebhookPolicy{
+				Service: nil,
+				URL:     &(server.URL),
+			},
+			specReplicas:            50,
+			statusReplicas:          50,
+			statusAllocatedReplicas: -20,
+			statusReadyReplicas:     40,
+			expected: expected{
+				replicas: 50,
+				limited:  false,
+				err:      "invalid character 'i' looking for beginning of value",
+			},
+		},
+		{
+			description: "Nil URL, empty namespace - default one should be used, no such host err is returned",
```

Comment:
I see that url is not nil here, cause it is `&url`

Reply:
URL as a WebhookClientConfig field is nil. It can be either URL or Service provided. Check this case, for example :https://github.com/googleforgames/agones/pull/1531/files/93d0ff777d37bc3a44357f02f56c7c3f40423ea4#diff-81c1d605645c546c9896631e7db55addR408